torch/csrc/utils/future.h

```diff
@@ -163,13 +153,19 @@ class TORCH_API Future final {
     value_ = std::move(value);
     completed_ = true;
 
+    executeCallbacksOnCompletion(lock);
+  }
+
+  // Runs all the callbacks attached to this future.
+  void executeCallbacksOnCompletion(std::unique_lock<std::mutex>& lock) {
```

Comment:
I think passing a ref to this lock around functions (this existed before this PR as well) could potentially get confusing, and it would be the developer's responsibility to unlock it properly. What about deduping the code into 2 portions:

1: setting error/value and completed, and moving the callbacks into a local vector
(unlock between these)
2: notifying on the CV and running end callbacks